client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java

```diff
@@ -427,52 +430,84 @@ private Properties loadFromStream(InputStream stream) {
-            TransactionMode transactionMode = getTransactionMode(transactionModeOverride);
-            RemoteCacheImpl<K, V> result;
-            if (transactionMode == TransactionMode.NONE) {
-               result = createRemoteCache(cacheName);
-            } else {
-               if (!this.isTransactional(cacheName)) {
-                  throw HOTROD.cacheDoesNotSupportTransactions(cacheName);
-               }
-               TransactionManager transactionManager = getTransactionManager(transactionManagerOverride);
-               result = createRemoteTransactionalCache(cacheName, forceReturnValueOverride,
-                     transactionMode == TransactionMode.FULL_XA, transactionMode, transactionManager);
-            }
-            RemoteCacheHolder rcc = new RemoteCacheHolder(result, forceReturnValueOverride);
-            startRemoteCache(rcc);
+      RemoteCacheConfiguration cacheConfiguration = configuration.remoteCaches().get(cacheName);
+      boolean forceReturnValue = forceReturnValueOverride ? true : (cacheConfiguration != null ? cacheConfiguration.forceReturnValues() : configuration.forceReturnValues());
+      RemoteCacheKey key = new RemoteCacheKey(cacheName, forceReturnValue);
+      if (cacheName2RemoteCache.containsKey(key)) {
+         return cacheName2RemoteCache.get(key).remoteCache();
+      }
 
-            PingResponse pingResponse = result.resolveStorage();
-            // If ping not successful assume that the cache does not exist
-            if (pingResponse.isCacheNotFound()) {
+      OperationsFactory operationsFactory = createOperationFactory(cacheName, forceReturnValue, codec, null);
+      PingResponse pingResponse;
+      if (started) {
+         // Verify if the cache exists on the server first
+         pingResponse = await(operationsFactory.newFaultTolerantPingOperation().execute());
+
+         // If ping not successful assume that the cache does not exist
+         if (pingResponse.isCacheNotFound()) {
+            // We may be able to create it. Don't use RemoteCacheAdmin for this, since it would end up calling this method again
+            Map<String, byte[]> params = new HashMap<>(2);
+            params.put(RemoteCacheManagerAdminImpl.CACHE_NAME, cacheName.getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            if (cacheConfiguration != null && cacheConfiguration.templateName() != null) {
+               params.put(RemoteCacheManagerAdminImpl.CACHE_TEMPLATE, cacheConfiguration.templateName().getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            } else if (cacheConfiguration != null && cacheConfiguration.configuration() != null) {
+               params.put(RemoteCacheManagerAdminImpl.CACHE_CONFIGURATION, new XMLStringConfiguration(cacheConfiguration.configuration()).toXMLString(cacheName).getBytes(HotRodConstants.HOTROD_STRING_CHARSET));
+            } else {
+               // We cannot create the cache
                return null;
```

Comment:
instead if returning null and exception or logging that would explain better the issue ?

Reply:
I really would like to throw an exception, but the previous version returned null and I didn't want to break backwards compat